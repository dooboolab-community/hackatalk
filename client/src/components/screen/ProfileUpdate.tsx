import * as Config from '../../../config';

import { Alert, Image, TouchableOpacity, View } from 'react-native';
import { Button, EditText } from 'dooboo-ui';
import { IC_CAMERA, IC_PROFILE } from '../../utils/Icons';
import React, { FC, useEffect, useState } from 'react';
import {
  fetchQuery,
  graphql,
  useRelayEnvironment,
} from 'react-relay/hooks';
import { launchCameraAsync, launchImageLibraryAsync } from '../../utils/ImagePicker';

import AsyncStorage from '@react-native-community/async-storage';
import { EditTextInputType } from 'dooboo-ui/EditText';
import { MainStackNavigationProps } from '../navigation/MainStackNavigator';
import type { ProfileUpdateMeQuery } from '../../__generated__/ProfileUpdateMeQuery.graphql';
import { getString } from '../../../STRINGS';
import { resizeImage } from '../../utils/image';
import styled from 'styled-components/native';
import { useActionSheet } from '@expo/react-native-action-sheet';
import { useThemeContext } from '@dooboo-ui/theme';

const BUTTON_INDEX_LAUNCH_CAMERA = 0;
const BUTTON_INDEX_LAUNCH_IMAGE_LIBLARY = 1;
const BUTTON_INDEX_CANCEL = 2;
const DEFAULT = {
  PROFILEIMAGE_WIDTH: 300,
  PROFILEIMAGE_HEIGHT: 300,
};

const Container = styled.View`
  flex: 1;
  background-color: ${({ theme }): string => theme.background};
  flex-direction: column;
  align-items: center;
`;

const StyledScrollView = styled.ScrollView`
  align-self: stretch;
`;

const Wrapper = styled.View`
  margin-top: 48px;
  width: 78%;
  flex-direction: column;
  align-items: center;
`;

const StyledButtonWrapper = styled.View`
  flex: 1;
  justify-content: space-between;
  flex-direction: row;
  align-items: center;
  align-self: stretch;
  height: 60px;
  margin-top: 36px;
  margin-bottom: 48px;
`;

const ProfileImage = styled.Image`
  width: 90px;
  height: 90px;
  border-radius: 45px;
`;

interface Props {
  navigation: MainStackNavigationProps<'ProfileUpdate'>;
}

const meQuery = graphql`
  query ProfileUpdateMeQuery {
    me {
      id
      email
      name
      nickname
      statusMessage
      verified
      profile {
        authType
      }
    }
  }
`;

const Screen: FC<Props> = () => {
  const { theme } = useThemeContext();
  const [name, setName] = useState('');
  const [nickname, setNickname] = useState('');
  const [statusMessage, setstatusMessage] = useState('');
  const { showActionSheetWithOptions } = useActionSheet();
  const [profilePath, setProfilePath] = useState('');
  const environment = useRelayEnvironment();

  useEffect(() => {
    fetchQuery<ProfileUpdateMeQuery>(environment, meQuery, {}).subscribe({
      next: (data) => {
        if (data.me) {
          const { name, nickname, statusMessage } = data.me;
          setName(name ?? '');
          setNickname(nickname ?? '');
          setstatusMessage(statusMessage ?? '');
        }
      },
    });
  }, []);

  const changeText = (type: string, text: string): void => {
    switch (type) {
      case 'NICKNAME':
        setNickname(text);
        break;
      case 'NAME':
        setName(text);
        break;
      case 'STATUS_MSG':
      default:
        setstatusMessage(text);
        break;
    }
  };

  const uploadImage = async (uri: string): Promise<string> => {
    const fileName = uri.split('/').pop() || '';
    const fileTypeMatch = /\.(\w+)$/.exec(fileName);
    const fileType = fileTypeMatch ? `image/${fileTypeMatch[1]}` : 'image';
    const data: FormData = new FormData();
    const token = await AsyncStorage.getItem('token');
    data.append('profile', {
      uri: uri,
      type: fileType,
      name: fileName,
    });

    const fetchInitOption = {
      method: 'POST',
      body: {
        inputFile: data,
        dir: 'profiles',
      },
      headers: new Headers({
        Accept: 'application/json',
        Authorization: `Bearer ${token}`,
      }),
    };

    return fetch(`${Config.ROOT_URL}/upload_single`, fetchInitOption)
      .then((res: Response) => res.url)
      .catch(() => {
        Alert.alert(getString('ERROR'), getString('ERROR_OCCURED'));
        return '';
      });
  };

  const pressProfileImage = async (): Promise<void> => {
    const options = [
      getString('TAKE_A_PICTURE'),
      getString('SELSCT_FROM_ALBUM'),
      getString('CANCEL'),
    ];

    showActionSheetWithOptions(
      {
        options,
        cancelButtonIndex: BUTTON_INDEX_CANCEL,
      },
      async (buttonIndex: number) => {
        if (buttonIndex === BUTTON_INDEX_LAUNCH_CAMERA) {
          const image = await launchCameraAsync();
          if (image && !image.cancelled) {
            const resizedImage = await resizeImage({
              imageUri: image.uri,
              maxWidth: DEFAULT.PROFILEIMAGE_WIDTH,
              maxHeight: DEFAULT.PROFILEIMAGE_HEIGHT,
            });
            const imageURL = await uploadImage(resizedImage.uri);

            setProfilePath(imageURL);
          }
          return;
        }

        if (buttonIndex === BUTTON_INDEX_LAUNCH_IMAGE_LIBLARY) {
          const image = await launchImageLibraryAsync();
          if (image && !image.cancelled) {
            const resizedImage = await resizeImage({
              imageUri: image.uri,
              maxWidth: DEFAULT.PROFILEIMAGE_WIDTH,
              maxHeight: DEFAULT.PROFILEIMAGE_HEIGHT,
            });
            const imageURL = await uploadImage(resizedImage.uri);

            setProfilePath(imageURL);
          }
        }
      },
    );
  };

  return (
    <Container>
      <StyledScrollView
        contentContainerStyle={{
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <Wrapper>
          <TouchableOpacity
            testID="button-user-icon"
            activeOpacity={0.5}
            onPress={pressProfileImage}
          >
            {!profilePath
              ? <View style={{
                width: 90,
                height: 90,
              }}>
                <Image style={{ height: 80, width: 80 }} source={IC_PROFILE} />
                <Image style={{
                  position: 'absolute',
                  bottom: 0,
                  right: 0,
                }} source={IC_CAMERA}/>
              </View>
              : <ProfileImage
                testID="profile-image"
                source={{ uri: profilePath }}
              />
            }
          </TouchableOpacity>
          <EditText
            testID="input-nickname"
            type={EditTextInputType.DEFAULT}
            style={{ marginTop: 32 }}
            textStyle={{ color: theme.fontColor }}
            label={getString('NICKNAME')}
            placeholder={getString('NICKNAME_HINT')}
            value={nickname}
            borderColor={theme.font}
            focusColor={theme.focused}
            placeholderTextColor={theme.placeholder}
            onChangeText={(text: string): void =>
              changeText('NICKNAME', text)
            }
          />
          <EditText
            testID="input-name"
            type={EditTextInputType.DEFAULT}
            style={{ marginTop: 32 }}
            textStyle={{ color: theme.fontColor }}
            label={getString('NAME')}
            placeholder={getString('NAME_HINT')}
            value={name}
            borderColor={theme.font}
            focusColor={theme.focused}
            placeholderTextColor={theme.placeholder}
            onChangeText={(text: string): void =>
              changeText('NAME', text)
            }
          />
          <EditText
            type={EditTextInputType.DEFAULT}
            testID="input-status"
            style={{ marginTop: 24 }}
            textStyle={{
              color: theme.fontColor,
            }}
            label={getString('STATUS_MSG')}
            placeholder={getString('STATUS_MSG_HINT')}
            value={statusMessage}
            borderColor={theme.font}
            focusColor={theme.focused}
            placeholderTextColor={theme.placeholder}
            onChangeText={(text: string): void =>
              changeText('STATUS_MSG', text)
            }
          />
          <StyledButtonWrapper>
            <Button
              testID="button-update"
              containerStyle={{ flex: 1 }}
              style={{
                width: '100%',
                backgroundColor: theme.btnPrimary,
                borderColor: theme.btnPrimary,
                borderWidth: 1,
              }}
              textStyle={{
                color: theme.btnPrimaryFont,
                fontSize: 14,
                fontWeight: 'bold',
              }}
              // isLoading={isUpdating}
              // onPress={updateProfile}
              text={getString('UPDATE')}
            />
          </StyledButtonWrapper>
        </Wrapper>
      </StyledScrollView>
    </Container>
  );
};

export default Screen;
