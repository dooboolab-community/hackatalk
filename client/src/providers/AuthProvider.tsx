import React, {useReducer} from 'react';

import {User} from '../types/graphql';
import createCtx from '../utils/createCtx';

export interface State {
  user: User | undefined;
}

interface Context {
  state: State;
  setUser(user: User | undefined): void;
}

const [useCtx, Provider] = createCtx<Context>();

// eslint-disable-next-line no-shadow
export enum ActionType {
  SetUser = 'set-user',
}

type SetUserAction = {
  type: ActionType.SetUser;
  payload: {user: User};
};
type Action = SetUserAction;

interface Props {
  children?: React.ReactElement;
  initialAuthUser?: User;
}

type Reducer = (state: State, action: Action) => State;

const setUser = (dispatch: React.Dispatch<SetUserAction>) => (
  authUser: User,
): void => {
  dispatch({
    type: ActionType.SetUser,
    payload: {user: authUser},
  });
};

const initialState: State = {
  user: undefined,
};

// eslint-disable-next-line default-param-last
const reducer: Reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SetUser:
      return {...action.payload};
    default:
      return state;
  }
};

function AuthProvider(props: Props): React.ReactElement {
  const initialAuthState = {
    user: props.initialAuthUser,
  };

  const [state, dispatch] = useReducer<Reducer>(reducer, initialAuthState);

  const actions = {
    setUser: setUser(dispatch),
  };

  return <Provider value={{state, ...actions}}>{props.children}</Provider>;
}

const AuthContext = {
  useAuthContext: useCtx,
  AuthProvider,
};

export {useCtx as useAuthContext, AuthProvider};
export default AuthContext;
