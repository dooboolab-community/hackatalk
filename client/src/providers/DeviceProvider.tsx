import * as Device from 'expo-device';

import React, {FC, useEffect, useState} from 'react';

import createCtx from '../utils/createCtx';

interface Context {
  deviceType: Device.DeviceType;
  setDeviceType: React.Dispatch<React.SetStateAction<Device.DeviceType>>;
}

const [useCtx, Provider] = createCtx<Context>();

const DeviceProvider: FC = ({children}) => {
  const [deviceType, setDeviceType] = useState<Device.DeviceType>(
    Device.DeviceType.PHONE,
  );

  useEffect(() => {
    let isMounted = true;

    Device.getDeviceTypeAsync().then((result) => {
      // Only change state if this component is still mounted.
      if (isMounted) setDeviceType(result);
    });

    return () => {
      isMounted = false;
    };
  }, []);

  return (
    <Provider
      value={{
        deviceType,
        setDeviceType,
      }}>
      {children}
    </Provider>
  );
};

const DeviceContext = {
  useDeviceContext: useCtx,
  DeviceProvider,
};

export {useCtx as useDeviceContext, DeviceProvider};
export default DeviceContext;
