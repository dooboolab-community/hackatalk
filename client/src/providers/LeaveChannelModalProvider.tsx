import React, {createContext, useContext, useState} from 'react';
import type {ReactElement, ReactNode} from 'react';

type LeaveChannelModalData = {
  testID?: string;
  channelId: string;
  onClickYes?: () => void;
};

export type LeaveChannelModalState =
  | {isVisible: false}
  | ({isVisible: true} & LeaveChannelModalData);

export type LeaveChannelModalContext = {
  showModal: (next: LeaveChannelModalData) => void;
  hideModal: () => void;
  modalState: LeaveChannelModalState;
};

const LeaveChannelModalContext = createContext<LeaveChannelModalContext>({
  modalState: {isVisible: false},
  showModal: () => {},
  hideModal: () => {},
});

type Props = {
  children: ReactNode;
};

export const LeaveChannelModalProvider = ({children}: Props): ReactElement => {
  const [showModalState, setModalState] = useState<LeaveChannelModalState>({
    isVisible: false,
  });

  const showModal = (next: LeaveChannelModalData): void =>
    setModalState({
      isVisible: true,
      ...next,
    });

  const hideModal = (): void =>
    setModalState({
      isVisible: false,
    });

  return (
    <LeaveChannelModalContext.Provider
      value={{
        showModal,
        hideModal,
        modalState: showModalState,
      }}
    >
      {children}
    </LeaveChannelModalContext.Provider>
  );
};

export const useLeaveChannelContext = (): LeaveChannelModalContext =>
  useContext(LeaveChannelModalContext);
