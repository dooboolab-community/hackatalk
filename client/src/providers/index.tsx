export * from './AuthProvider';
export * from './DeviceProvider';
export * from './ProfileModalProvider';
