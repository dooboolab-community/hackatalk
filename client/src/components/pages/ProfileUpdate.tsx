import {Alert, Image, TouchableOpacity, View} from 'react-native';
import {Button, EditText, useTheme} from 'dooboo-ui';
import {IC_CAMERA, IC_PROFILE} from '../../utils/Icons';
import React, {FC, useCallback, useEffect, useRef, useState} from 'react';
import {
  UploadSingleUploadMutation,
  UploadSingleUploadMutationResponse,
} from '../../__generated__/UploadSingleUploadMutation.graphql';
import {fetchQuery, useMutation, useRelayEnvironment} from 'react-relay';
import {
  launchCameraAsync,
  launchImageLibraryAsync,
} from '../../utils/ImagePicker';
import {meQuery, profileUpdate} from '../../relay/queries/User';

import {ImagePickerResult} from 'expo-image-picker';
import {ReactNativeFile} from 'apollo-upload-client';
import type {UserMeQuery} from '../../__generated__/UserMeQuery.graphql';
import type {UserUpdateProfileMutation} from '../../__generated__/UserUpdateProfileMutation.graphql';
import {getString} from '../../../STRINGS';
import {resizePhotoToMaxDimensionsAndCompressAsPNG} from '../../utils/image';
import {showAlertForError} from '../../utils/common';
import {singleUpload} from '../../relay/queries/Upload';
import styled from '@emotion/native';
// import {uploadImageAsync} from '../../apis/upload';
import {useActionSheet} from '@expo/react-native-action-sheet';
import {useAuthContext} from '../../providers/AuthProvider';

const BUTTON_INDEX_LAUNCH_CAMERA = 0;
const BUTTON_INDEX_LAUNCH_IMAGE_LIBRARY = 1;
const BUTTON_INDEX_CANCEL = 2;

const DEFAULT = {
  PROFILEIMAGE_WIDTH: 1280,
  PROFILEIMAGE_HEIGHT: 1280,
};

const Container = styled.View`
  flex: 1;
  background-color: ${({theme}) => theme.background};
  flex-direction: column;
  align-items: center;
`;

const StyledScrollView = styled.ScrollView`
  align-self: stretch;
`;

const Wrapper = styled.View`
  margin-top: 48px;
  width: 78%;
  flex-direction: column;
  align-items: center;
`;

const StyledButtonWrapper = styled.View`
  flex: 1;
  justify-content: space-between;
  flex-direction: row;
  align-items: center;
  align-self: stretch;
  height: 60px;
  margin-top: 36px;
  margin-bottom: 48px;
`;

const ProfileImage = styled.Image`
  width: 90px;
  height: 90px;
  border-radius: 45px;
`;

const Screen: FC = () => {
  const {theme} = useTheme();
  const [name, setName] = useState('');
  const [nickname, setNickname] = useState('');
  const [statusMessage, setstatusMessage] = useState('');
  const {showActionSheetWithOptions} = useActionSheet();
  const [profilePath, setProfilePath] = useState('');
  const environment = useRelayEnvironment();
  const envrionmentProps = useRef(environment);

  const {
    state: {user},
  } = useAuthContext();

  const [
    commitUpload,
    isUploadInFlight,
  ] = useMutation<UploadSingleUploadMutation>(singleUpload);

  const [
    commitProfileUpdate,
    isUpdating,
  ] = useMutation<UserUpdateProfileMutation>(profileUpdate);

  const [isUploading, setIsUploading] = useState(false);

  useEffect(() => {
    fetchQuery<UserMeQuery>(envrionmentProps.current, meQuery, {}).subscribe({
      next: (data) => {
        if (data.me) {
          const {
            name: myName,
            nickname: nickName,
            statusMessage: statusMsg,
            photoURL,
            thumbURL,
          } = data.me;

          setName(myName ?? '');
          setNickname(nickName ?? '');
          setstatusMessage(statusMsg ?? '');
          setProfilePath((thumbURL || photoURL) ?? '');
        }
      },
    });
  }, []);

  const changeText = (type: string, text: string): void => {
    switch (type) {
      case 'NICKNAME':
        setNickname(text);
        break;
      case 'NAME':
        setName(text);
        break;
      case 'STATUS_MSG':
      default:
        setstatusMessage(text);
        break;
    }
  };

  const processImageUpload = useCallback(
    async (image: ImagePickerResult | null): Promise<void> => {
      if (image && !image.cancelled) {
        setIsUploading(true);

        try {
          const resizedImage = await resizePhotoToMaxDimensionsAndCompressAsPNG(
            {
              uri: image.uri,
              width: DEFAULT.PROFILEIMAGE_WIDTH,
              height: DEFAULT.PROFILEIMAGE_HEIGHT,
            },
          );

          const fileName = resizedImage.uri.split('/').pop() || '';
          const fileTypeMatch = /\.(\w+)$/.exec(fileName);

          const fileType = fileTypeMatch
            ? `image/${fileTypeMatch[1]}`
            : 'image';

          const file = new ReactNativeFile({
            uri: resizedImage.uri,
            name: `${user?.id || fileName}.${fileType}`,
            type: fileType,
          });

          commitUpload({
            variables: {
              file: null,
              dir: 'profiles',
            },
            uploadables: {
              file,
            },
            onCompleted: (response: UploadSingleUploadMutationResponse) => {
              const url = response.singleUpload;

              if (url) setProfilePath(url);

              setIsUploading(false);
            },
            onError: (err) => {
              Alert.alert(getString('ERROR'), err.message);
            },
          });
        } catch (err) {
          Alert.alert(getString('ERROR'), getString('FAILED_LOAD_IMAGE'));
        } finally {
          setIsUploading(false);
        }
      }
    },
    [commitUpload, user?.id],
  );

  const pressProfileImage = async (): Promise<void> => {
    const options = [
      getString('TAKE_A_PICTURE'),
      getString('SELECT_FROM_ALBUM'),
      getString('CANCEL'),
    ];

    showActionSheetWithOptions(
      {
        options,
        cancelButtonIndex: BUTTON_INDEX_CANCEL,
      },
      async (buttonIndex: number) => {
        if (buttonIndex === BUTTON_INDEX_LAUNCH_CAMERA) {
          const image = await launchCameraAsync();

          processImageUpload(image);

          return;
        }

        if (buttonIndex === BUTTON_INDEX_LAUNCH_IMAGE_LIBRARY) {
          const image = await launchImageLibraryAsync();

          processImageUpload(image);
        }
      },
    );
  };

  const updateProfile = async (): Promise<void> => {
    const mutationConfig = {
      variables: {
        user: {
          name,
          nickname,
          statusMessage,
          thumbURL: profilePath,
          photoURL: profilePath,
        },
      },
      onError: (error: Error): void => {
        showAlertForError(error);
      },
    };

    commitProfileUpdate(mutationConfig);
  };

  return (
    <Container>
      <StyledScrollView
        contentContainerStyle={{
          justifyContent: 'center',
          alignItems: 'center',
        }}>
        <Wrapper>
          <TouchableOpacity
            testID="button-user-icon"
            activeOpacity={0.5}
            onPress={pressProfileImage}>
            {!profilePath ? (
              <View
                style={{
                  width: 90,
                  height: 90,
                }}>
                <Image style={{height: 80, width: 80}} source={IC_PROFILE} />
                <Image
                  style={{
                    position: 'absolute',
                    bottom: 0,
                    right: 0,
                  }}
                  source={IC_CAMERA}
                />
              </View>
            ) : (
              <ProfileImage
                testID="profile-image"
                resizeMode="cover"
                source={{uri: profilePath}}
              />
            )}
          </TouchableOpacity>
          <EditText
            testID="input-nickname"
            style={{marginTop: 32}}
            styles={{
              container: {
                borderColor: theme.text,
              },
              input: {
                color: theme.text,
              },
            }}
            labelText={getString('NICKNAME')}
            placeholder={getString('NICKNAME_HINT')}
            value={nickname}
            focusColor={theme.focused}
            placeholderTextColor={theme.placeholder}
            onChangeText={(text: string): void => changeText('NICKNAME', text)}
          />
          <EditText
            testID="input-name"
            style={{marginTop: 32}}
            styles={{
              container: {
                borderColor: theme.text,
              },
              input: {
                color: theme.text,
              },
            }}
            labelText={getString('NAME')}
            placeholder={getString('NAME_HINT')}
            value={name}
            focusColor={theme.focused}
            placeholderTextColor={theme.placeholder}
            onChangeText={(text: string): void => changeText('NAME', text)}
          />
          <EditText
            testID="input-status"
            type="column"
            style={{marginTop: 36}}
            styles={{
              input: {
                marginTop: 12,
                color: theme.text,
              },
              container: {
                borderColor: theme.text,
                borderWidth: 1,
                paddingHorizontal: 8,
                paddingVertical: 12,
              },
            }}
            labelText={getString('STATUS_MSG')}
            placeholder={getString('STATUS_MSG_HINT')}
            value={statusMessage}
            focusColor={theme.focused}
            placeholderTextColor={theme.placeholder}
            onChangeText={(text: string): void =>
              changeText('STATUS_MSG', text)
            }
            textInputProps={{
              multiline: true,
            }}
          />
          <StyledButtonWrapper>
            <Button
              testID="button-update"
              style={{alignSelf: 'stretch', flex: 1}}
              styles={{
                container: {
                  height: 44,
                  backgroundColor: theme.btnPrimary,
                  borderColor: theme.btnPrimary,
                  borderWidth: 1,
                },
                text: {
                  color: theme.btnPrimaryFont,
                  fontSize: 14,
                  fontWeight: 'bold',
                },
              }}
              loading={isUploading || isUpdating || isUploadInFlight}
              onPress={updateProfile}
              text={getString('UPDATE')}
            />
          </StyledButtonWrapper>
        </Wrapper>
      </StyledScrollView>
    </Container>
  );
};

export default Screen;
