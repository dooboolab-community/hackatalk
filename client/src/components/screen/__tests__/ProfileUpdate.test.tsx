import 'react-native';

import React, { ReactElement } from 'react';
import {
  cleanup,
  fireEvent,
  render,
  waitFor,
} from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import { MockPayloadGenerator } from 'relay-test-utils';
import ProfileUpdate from '../ProfileUpdate';
import ProfileUpdateMeQuery from '../../../__generated__/ProfileUpdateMeQuery.graphql';
import { environment } from '../../../providers';

const component = createTestElement(
  <ProfileUpdate {...createTestProps()} />,
);

jest.mock('@expo/react-native-action-sheet', () => ({
  useActionSheet: (): Record<string, unknown> => {
    const userPressLaunchCamera = true;
    const BUTTON_INDEX_LAUNCH_CAMERA = 0;
    const BUTTON_INDEX_LAUNCH_IMAGE_LIBRARY = 1;

    return {
      showActionSheetWithOptions: (
        options: Record<string, unknown>,
        callback: (index: number) => void,
      ): void => {
        if (userPressLaunchCamera) {
          callback(BUTTON_INDEX_LAUNCH_CAMERA);
        }

        callback(BUTTON_INDEX_LAUNCH_IMAGE_LIBRARY);
      },
    };
  },
}));

jest.mock('expo-permissions', () => ({
  askAsync: (): { status: string } => ({
    status: 'granted',
  }),
}));

jest.mock('expo-image-picker', () => ({
  launchCameraAsync: (): string => 'photo info',
  launchImageLibraryAsync: (): string => 'photo info',
}));

describe('rendering test', () => {
  it('renders as expected', () => {
    const mockResolver = {
      Profile: () => ({
        id: 'id',
        email: 'email',
        name: 'name',
        nickname: 'nickname',
        statusMessage: '',
        verified: true,
      }),
    };

    environment.mock.queueOperationResolver(
      (operation) => MockPayloadGenerator.generate(operation, mockResolver),
    );

    const json = render(component).toJSON();

    expect(json).toBeTruthy();
    expect(json).toMatchSnapshot();
  });
});

describe('[ProfileUpdate] interaction', () => {
  it('should change nickname', async () => {
    const { getByTestId } = render(component);
    const inputStatus = getByTestId('input-nickname');

    await waitFor(() => expect(inputStatus).toBeTruthy());

    fireEvent.changeText(inputStatus, 'nickname');

    expect(inputStatus.props.value).toEqual('nickname');
  });

  it('should change name', async () => {
    const { getByTestId } = render(component);
    const inputName = getByTestId('input-name');

    await waitFor(() => expect(inputName).toBeTruthy());

    fireEvent.changeText(inputName, 'name');

    expect(inputName.props.value).toEqual('name');
  });

  it('should change status text', async () => {
    const { getByTestId } = render(component);
    const inputStatus = getByTestId('input-status');

    await waitFor(() => expect(inputStatus).toBeTruthy());

    fireEvent.changeText(inputStatus, 'status');

    expect(inputStatus.props.value).toEqual('status');
  });

  it('should call updateProfile when update button pressed', async () => {
    const { getByTestId } = render(component);
    const updateButton = getByTestId('button-update');

    await waitFor(() => expect(updateButton).toBeTruthy());

    fireEvent.press(updateButton);

    const operation = environment.mock.getMostRecentOperation();

    environment.mock.resolve(
      operation,
      MockPayloadGenerator.generate(operation),
    );
  });

  // it('should launch camera when user select "Take a picture"', async () => {
  //   // userPressLaunchCamera = true;
  //   const profileBtn = await waitForElement(() =>
  //     testingLib.queryByTestId('button-user-icon'),
  //   );
  //   await act(() => {
  //     fireEvent.press(profileBtn);
  //     // jest.runAllTimers();
  //   });
  //   await wait();
  // });

  // it('should open album when user select "Select from Album"', async () => {
  //   // userPressLaunchCamera = false;
  //   const profileBtn = await waitForElement(() =>
  //     testingLib.queryByTestId('button-user-icon'),
  //   );
  //   await wait(() => expect(profileBtn).toBeTruthy());
  //   act(() => {
  //     fireEvent.press(profileBtn);
  //     // jest.runAllTimers();
  //   });
  //   await wait();
  // });
});
