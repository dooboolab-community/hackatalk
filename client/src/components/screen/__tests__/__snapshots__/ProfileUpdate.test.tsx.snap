// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering test renders as expected 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    style={
      Array [
        Object {
          "alignSelf": "stretch",
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "marginTop": 48,
              "width": "78%",
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="button-user-icon"
        >
          <View
            style={
              Object {
                "height": 90,
                "width": 90,
              }
            }
          >
            <Image
              source={
                Object {
                  "canInstrument": true,
                  "createTransformer": [Function],
                  "getCacheKey": [Function],
                  "process": [Function],
                }
              }
              style={
                Object {
                  "height": 80,
                  "width": 80,
                }
              }
            />
            <Image
              source={
                Object {
                  "canInstrument": true,
                  "createTransformer": [Function],
                  "getCacheKey": [Function],
                  "process": [Function],
                }
              }
              style={
                Object {
                  "bottom": 0,
                  "position": "absolute",
                  "right": 0,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "flexDirection": "column",
              },
              Object {
                "marginTop": 32,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#b9b9c4",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "marginBottom": 5,
                },
                undefined,
                false,
              ]
            }
          >
            Nickname
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="A Nickname can help others find you"
            placeholderTextColor="#676c7a"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "fontWeight": "500",
                  "minHeight": 44,
                  "paddingBottom": 15,
                  "paddingTop": 15,
                },
                Object {
                  "color": "#151A25",
                },
              ]
            }
            testID="input-nickname"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
                undefined,
                Object {
                  "borderBottomWidth": 0.6,
                  "borderColor": "#2E2E2E",
                },
                false,
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "flexDirection": "column",
              },
              Object {
                "marginTop": 32,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#b9b9c4",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "marginBottom": 5,
                },
                undefined,
                false,
              ]
            }
          >
            Name
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Please write your name"
            placeholderTextColor="#676c7a"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "fontWeight": "500",
                  "minHeight": 44,
                  "paddingBottom": 15,
                  "paddingTop": 15,
                },
                Object {
                  "color": "#151A25",
                },
              ]
            }
            testID="input-name"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
                undefined,
                Object {
                  "borderBottomWidth": 0.6,
                  "borderColor": "#2E2E2E",
                },
                false,
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "flexDirection": "column",
              },
              Object {
                "marginTop": 24,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#b9b9c4",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "marginBottom": 5,
                },
                undefined,
                false,
              ]
            }
          >
            Status message
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Please write your status message"
            placeholderTextColor="#676c7a"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "fontWeight": "500",
                  "minHeight": 44,
                  "paddingBottom": 15,
                  "paddingTop": 15,
                },
                Object {
                  "color": "#151A25",
                },
              ]
            }
            testID="input-status"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
                undefined,
                Object {
                  "borderBottomWidth": 0.6,
                  "borderColor": "#2E2E2E",
                },
                false,
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 1,
                "flexShrink": 1,
                "height": 60,
                "justifyContent": "space-between",
                "marginBottom": 48,
                "marginTop": 36,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#1E6EFA",
                "borderColor": "#1E6EFA",
                "borderWidth": 1,
                "opacity": 1,
                "width": "100%",
              }
            }
            testID="button-update"
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
